### PR TITLE
Add support for multiple configs to TuistGenerator

### PR DESCRIPTION
**Pull request by marciniwanicki**
Part of https://github.com/tuist/tuist/issues/160

### Short description 📝

As mentioned in https://github.com/tuist/tuist/issues/160, currently, Tuist only supports debug and release build configurations. This can be seen as a limitation for some projects. We'd like to avoid this limitation and allow users to freely control the configurations themselves and base them on xcconfig files if needed.

### Solution 📦

Added `Settings` property to `Project` and `Target` generator models. `Settings` contains `base` property for settings that are common for all the configurations. `configurations` property stores dictionary mapping `BuildConfiguration` (i.e. Release, Debug, Beta etc.) to `Configuration`. The structure won't preserve the order of the configurations but can help avoid situation when the same configuration is set twice (which would require additional linting). Maintaining the original order would be particularly difficult in cases when the defined configurations orders are different in project and target settings. 

i.e.
```swift
Project(settings: Settings(configurations: [Configuration(.debug), Configuration(.release))
Target(settings: Settings(configurations: [Configuration(.release), Configuration(.debug))
```

To keep the ordering stable, the generator sort the configurations alphabetically. 

`Configuration` contains settings (key-value pairs) and optionally a link to `.xcconfig` file.

```swift
public class Configuration: Equatable {
    public let settings: [String: String]
    public let xcconfig: AbsolutePath? 
    ...
}

public class Settings: Equatable {
    public let base: [String: String]
    public let configurations: [BuildConfiguration: Configuration?]
    ...
}
```

To make the implementation backward compatible, `GeneratorModelLoader` always passes release and debug configurations with the settings defined in the manifest.
```swift
return TuistGenerator.Settings(base: base, configurations: [.debug: debug, .release: release]
```

This issue won't solve https://github.com/tuist/tuist/issues/160 but it can be a good first step forward. I aimed to make the change fully transparent for the existing users. 
I'd be happy to create a separated PR to add `Settings` to `Project` and `Target` manifest types mapping almost one to one to Generator model `Settings` type. 
Using just settings properties could be inconvenient (quite a lot of copy-paste boilerplate) for bigger projects but can help smaller projects to leverage multiple configurations without waiting for "sharing" feature (Environments, Shared Configurations etc).

### CODE REVIEW

**Comment by marciniwanicki** [comment]
I think the best way to test the changes is using a dummy auto-generated project with different configurations, then running generation (via the library) and verifying the `xcodebuild -showBuildSettings` output.
I've created few tests here: https://github.com/bloomberg/tuist/tree/feature/160-generator-support-multiple-configs-integration-tests

Not much, it's just one extra file https://github.com/bloomberg/tuist/blob/feature/160-generator-support-multiple-configs-integration-tests/Tests/TuistGeneratorTests/Integration/MultipleConfigurationsIntegrationTests.swift

The caveat is that it will visibly slow down the unit tests. On my laptop I need ~4s for `swift test` on master. Running just `MultipleConfigurationsIntegrationTests.swift` takes 12s :( 

I was trying to find a better way to add those tests, but as long as SwiftPM does not support custom schemes, it's not that straightforward. What do you think guys? Do you have some ideas how we could integrate those tests?

File: Sources/TuistGenerator/Models/BuildConfiguration.swift [hunk 1]
```diff
@@ -1,15 +1,50 @@
...
+        case debug, release
+    }
+
+    public static let debug = BuildConfiguration(name: "Debug", predefined: true, variant: .debug)
+    public static let release = BuildConfiguration(name: "Release", predefined: true, variant: .release)
+
+    public let name: String
+    public let variant: Variant
+
+    private let predefined: Bool
```
Comment (pepibumur) [inline Sources/TuistGenerator/Models/BuildConfiguration.swift]: What's this attribute used for?

File: Sources/TuistGenerator/Models/Settings.swift [hunk 2]
```diff
@@ -29,27 +23,43 @@ public class Configuration: Equatable {
...
 
 public class Settings: Equatable {
+    public static let `default` = Settings(configurations: [.release: nil, .debug: nil])
+
     // MARK: - Attributes
 
     public let base: [String: String]
-    public let debug: Configuration?
-    public let release: Configuration?
+    public let configurations: [BuildConfiguration: Configuration?]
```
Comment (pepibumur) [inline Sources/TuistGenerator/Models/Settings.swift]: 👏

File: Sources/TuistGenerator/Models/Settings.swift [hunk 3]
```diff
@@ -29,27 +23,43 @@ public class Configuration: Equatable {
...
+                configurations: [BuildConfiguration: Configuration?]) {
         self.base = base
-        self.debug = debug
-        self.release = release
+        self.configurations = configurations
+    }
+
+    // MARK: - Internal
+
+    static func ordered(configurations: [BuildConfiguration: Configuration?])
```
Comment (pepibumur) [inline Sources/TuistGenerator/Models/Settings.swift]: Nitpick: What about moving this one into an extension to hashes whose key is a `BuildConfiguration` and the value is a `Configuration`?

Comment (pepibumur) [reply Sources/TuistGenerator/Models/Settings.swift]: Same comment applies to the two methods below 😛. It'll be easier to unit-test them.

**Comment by kwridan** [comment]
Any thoughts on the integration tests @marciniwanicki wrote? They do add time ⏳ to the tests, but provide the most confidence ✅

File: Sources/TuistGenerator/Models/Settings.swift (hunk 3 above)
Comment (marciniwanicki) [reply Sources/TuistGenerator/Models/Settings.swift]: Oh, right! Extension seems to fit a way better 👍. I wasn't sure where to put the tests for the extension methods, and they ended up inside `SettingsTests` suite. Would you prefer to use `DictionaryBuildConfigurationExtensionTests` or something similar?

File: Sources/TuistGenerator/Models/BuildConfiguration.swift (hunk 1 above)
Comment (marciniwanicki) [reply Sources/TuistGenerator/Models/BuildConfiguration.swift]: I put it only for debugging purposes, it does not play any role except this tint bit of extra information if the Configuration was created "manually" by the client code or it's one of the 2 predefined (`.debug` and `.release`) - that's it. I am happy to remove it.

**Comment by pedropinerashopify** [comment]
> Any thoughts on the [integration tests](https://github.com/tuist/tuist/pull/320#issuecomment-481774652) @marciniwanicki wrote? They do add time ⏳ to the tests, but provide the most confidence ✅

I like the approach of the test using `xcodebuild`'s `showBuildSettings` argument. However, I would not mix integration test with unit tests. Besides taking more time, they might add flakiness to the existing test suite.

Since defining those tests with Cucumber would be a bit inconvenient because we'd need a fixture for each scenario that you are testing, I'd create a new target on the Swift Package, `TuistIntegrationTests` and add those tests there.

File: Sources/TuistGenerator/Generator/ConfigGenerator.swift [hunk 4]
```diff
@@ -66,23 +64,29 @@ final class ConfigGenerator: ConfigGenerating {
...
-                                      graph: graph,
-                                      pbxproj: pbxproj,
-                                      configurationList: configurationList,
-                                      sourceRootPath: sourceRootPath)
+        let projectBuildConfigurations = projectSettings.configurations.keys
+        let targetConfigurations = target.settings?.configurations ?? [:]
+        let targetBuildConfigurations = targetConfigurations.keys
+        let buildConfigurations = Set(projectBuildConfigurations).union(targetBuildConfigurations)
+        let configurationsTuples: [(BuildConfiguration, Configuration?)] = buildConfigurations.map {
+            if let configuration = target.settings?.configurations[$0] {
```
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/ConfigGenerator.swift]: Nitpick: It's hard to guess what `$0` refers to here. What about naming the closure argument?

**Pull request merged**